## Heroes of Pymoli Data Analysis

In [ ]:
import pandas as pd

In [ ]:
pymoli = "purchase_data.json"

In [149]:
#test the table
data = pd.read_json(pymoli)
data.head()

Age Gender  Item ID                                  Item Name  Price  \
0   38   Male      165                Bone Crushing Silver Skewer   3.37   
1   21   Male      119  Stormbringer, Dark Blade of Ending Misery   2.32   
2   34   Male      174                            Primitive Blade   2.46   
3   21   Male       92                               Final Critic   1.36   
4   23   Male       63                             Stormfury Mace   1.27   

             SN  
0     Aelalis34  
1        Eolo46  
2   Assastnya25  
3  Pheusrical25  
4        Aela59

In [78]:
#test any NA data or empty rows
data.count()

Age          780
Gender       780
Item ID      780
Item Name    780
Price        780
SN           780
dtype: int64

In [79]:
#check data types
data.dtypes

Age            int64
Gender        object
Item ID        int64
Item Name     object
Price        float64
SN            object
dtype: object

In [80]:
#total number of players
data_players = len(data["SN"].unique())
print("Total number of Heroes of Pymoli players:", data_players)

Total number of Heroes of Pymoli players: 573


In [152]:
#Purchasing Analysis (Total): number of uniqe items, avg purchase price, total number of purchases, total revenue
#number of unique items
data_items = len(data["Item ID"].value_counts())
#total revenue 
data_revenue = data["Price"].sum()
#total number of purchases
data_itemcount = data["Price"].count()
#avg purchase price
data_avg = round(data_revenue/data_itemcount, 2)
#data frame of purchasing analysis
summary_purchase = pd.DataFrame({"Number of Unique Items": [data_items],
                         "Average purchase price": [data_avg],
                         "Total revenue($)": [data_revenue],
                         "Total number of purchase": [data_itemcount]
                                })
print("***PURCHASE ANALYSIS***")
summary_purchase

***PURCHASE ANALYSIS***


Number of Unique Items  Average purchase price  Total revenue($)  \
0                     183                    2.93           2286.33   

   Total number of purchase  
0                       780

In [82]:
#Before we are making gender demographic analysis, we have to go over each data and identify any differences
#There is a data conflict where total number of players is 573, but the total number of male/female/other count is 780 as it is shown in table below.
#it means some players with same ID made a mutiple purchases
#therefore, we have to aim for the total number 573 in order to make an accurate gender demographic analysis
data_gender = data["Gender"].value_counts()
data_gender

Male                     633
Female                   136
Other / Non-Disclosed     11
Name: Gender, dtype: int64

In [83]:
#first, we are going to discover any duplicate purchase on one account
data["SN"].value_counts()

Undirrala66        5
Saedue76           4
Qarwen67           4
Mindimnya67        4
Sondastan54        4
Hailaphos89        4
Mindirra92         3
Haerith37          3
Yasriphos60        3
Yadanun74          3
Ila44              3
Seorithstilis90    3
Isurria36          3
Frichosiala98      3
Tillyrin30         3
Chanastsda67       3
Sondim43           3
Lirtosia72         3
Aerithllora36      3
Eoda93             3
Lisistaya47        3
Ialistidru50       3
Chamistast30       3
Eusri70            3
Lisossa63          3
Chadjask77         3
Isursti83          3
Chadossa56         3
Aeduera68          3
Frichaya88         3
                  ..
Ililsan66          1
Baelollodeu94      1
Lirtyrdesta65      1
Pheusrical25       1
Jiskimsda56        1
Lisassa26          1
Aesty51            1
Undistasta86       1
Aillyriadru65      1
Alim85             1
Chanirra56         1
Quarunarn52        1
Yalostiphos68      1
Quelaton80         1
Iallyphos37        1
Saelollop56        1
Eullydru35   

In [84]:
#Then, we will make a group by SN then find out the gender distribution
#this is a list of each players' purchses based on their gender distribution
#at the end, we can confirm that it matches with the total number of player:573
group_gender = data.groupby(["Gender", "SN"])
group_SN = pd.DataFrame(group_gender["Price"].sum())
group_SN.head

<bound method NDFrame.head of                                       Price
Gender                SN                   
Female                Aenarap34        1.65
                      Aeral43          2.72
                      Aeri84           6.60
                      Aiduesu83        2.63
                      Ailaesuir66      1.88
                      Aina42           3.01
                      Aina43           2.46
                      Airithrin43      3.57
                      Aisur51          1.49
                      Alaesu91         1.03
                      Alim85           3.61
                      Anallorgue57     4.87
                      Assassasta79     9.12
                      Chadossa56       7.85
                      Chamadar27       4.23
                      Chamilsan75      5.64
                      Chamistast30     4.49
                      Chanastnya43     1.36
                      Chanosiast43     5.92
                      Chanosseya79     4.00
  

In [85]:
#now we can count the final number of male/female/other distribtuion within SN(unique ID)
gender_count = group_SN.count(level ="Gender")
count_rename = gender_count.rename(columns={"Price": "Gender Count"})
count_rename

Gender Count
Gender                             
Female                          100
Male                            465
Other / Non-Disclosed             8

In [150]:
#gender Demographics
#percentage/count of gender 
female_perct = round(100/573*100, 2)
male_perct = round(465/573*100, 2)
other_perct = round(8/573*100, 2)
gender_perct = [female_perct,male_perct,other_perct]
count_rename["Gender Percentage"] = gender_perct
count_rename["Gender Percentage"] = count_rename["Gender Percentage"].map("{:.2f}%".format)


In [87]:
#purchase count, avg purchase, total purchase, normalized totals
#first make female data table
female_data = data.loc[data["Gender"] == "Female", :]
#female purchse count
female_buy = female_data["Price"].count()
#female purchase total
female_sum = round((female_data["Price"].sum()),2)
#female avg
female_avg = round((female_data["Price"].mean()),2)
#female normalized total
female_normal = round((female_sum/female_buy),2)

In [88]:
#first make male data table
male_data = data.loc[data["Gender"] == "Male", :]
#fmale purchse count
male_buy = male_data["Price"].count()
#male purchase total
male_sum = round((male_data["Price"].sum()),2)
#male avg
male_avg = round((male_data["Price"].mean()),2)
#male normalized total
male_normal = round((male_sum/male_buy),2)

In [89]:
#first make other data table
other_data = data.loc[data["Gender"] == "Other / Non-Disclosed", :]
#fmale purchse count
other_buy = other_data["Price"].count()
#male purchase total
other_sum = round((other_data["Price"].sum()),2)
#male avg
other_avg = round((other_data["Price"].mean()),2)
#male normalized total
other_normal = round((other_sum/other_buy),2)

In [90]:
#comnaing data to make a table
gender_buy = [female_buy, male_buy, other_buy]
gender_sum = [female_sum, male_sum, other_sum]
gender_avg = [female_avg, male_avg, other_avg]
gender_normal = [female_normal, male_normal, other_normal]

In [151]:
#Final Gender Analysis
count_rename["Gender Purchase Count"] = gender_buy
count_rename["Gender Purchase Total"] = gender_sum
count_rename["Gender Purchase Total"] = count_rename["Gender Purchase Total"].map("${:.2f}".format)
count_rename["Gender Purchase Average"] = gender_avg
count_rename["Gender Purchase Average"] = count_rename["Gender Purchase Average"].map("${:.2f}".format)
count_rename["Gender Purchase Normalized Total"] = gender_normal
count_rename["Gender Purchase Normalized Total"] = count_rename["Gender Purchase Normalized Total"].map("${:.2f}".format)
print("***GENDER DEMOGRAPHICS***")
count_rename

***GENDER DEMOGRAPHICS***


Gender Count Gender Percentage  Gender Purchase Count  \
Gender                                                                         
Female                          100            17.45%                    136   
Male                            465            81.15%                    633   
Other / Non-Disclosed             8             1.40%                     11   

                      Gender Purchase Total Gender Purchase Average  \
Gender                                                                
Female                              $382.91                   $2.82   
Male                               $1867.68                   $2.95   
Other / Non-Disclosed                $35.74                   $3.25   

                      Gender Purchase Normalized Total  
Gender                                                  
Female                                           $2.82  
Male                                             $2.95  
Other / Non-Disclosed                            $3.25

In [92]:
#Age Demographics**
#The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.) 
#Purchase Count
#Average Purchase Price
#Total Purchase Value
#Normalized Totals

In [93]:
#first find out the min and max age numbers
print(data["Age"].min())
print(data["Age"].max())

7
45


In [144]:
#create age-group table as same as gender analysis table method
bins = [0, 10, 14, 18, 22, 26, 30, 34, 38, 42, 46]
group = ["<10", "10-14", "14-18", "18-22", "22-26", "26-30", "30-34","34-38", "38-42", "42-46"]
data["Age Group"] = pd.cut(data["Age"], bins, labels=group)
group_age = data.groupby(["Age Group"])
age_group = pd.DataFrame(group_age["Price"].sum())
age_rename = age_group.rename(columns={"Price": "Total Purchase"})
age_count = pd.DataFrame(group_age["Price"].count())
age_count["Total Purchase"] = age_rename
age_count_rename = age_count.rename(columns={"Price": "Purchase Count"})
#check the table and check the sum of Total Purchase mathes with the Total Revenue number
#calculate avg and normalized totals
age_avg = pd.DataFrame(round(group_age["Price"].mean(),2))
age_count_rename["Average Purchase"] = age_avg

In [145]:
#in order to calculate the count of each age groups, we have to eliminate any duplicated players ID who made a multiple purchases
#as a result, total number should be 573 as below table.
no_duplicate = data.drop_duplicates(subset = "SN", keep='first')
age_duplicate = no_duplicate["Age Group"].value_counts()
age_duplicate

18-22    178
22-26    153
14-18     84
26-30     44
30-34     34
34-38     25
<10       22
10-14     20
38-42     11
42-46      2
Name: Age Group, dtype: int64

In [147]:
#age group analysis
age_count_rename["Age Count"] = age_duplicate
age_perct = round(age_duplicate/573*100, 2)
age_count_rename["Age Percentitle"] = age_perct
age_normal = round(pd.DataFrame(age_count_rename["Total Purchase"]/age_count_rename["Age Count"]),2)
age_count_rename["Normalized Purchase"] = age_normal
age_count_rename ["Age Percentitle"] = age_count_rename["Age Percentitle"].map("{:.2f}%".format)
age_count_rename["Total Purchase"] = age_count_rename["Total Purchase"].map("${:.2f}".format)
age_count_rename["Average Purchase"] = age_count_rename["Average Purchase"].map("${:.2f}".format)
age_count_rename["Normalized Purchase"] = age_count_rename["Normalized Purchase"].map("${:.2f}".format)

In [148]:
#Final age analysis table
age_summary = age_count_rename[["Age Count", "Age Percentitle", "Total Purchase", "Purchase Count", "Average Purchase", "Normalized Purchase"]]
print("***AGE GROUP ANALYSIS***")
age_summary

***AGE GROUP ANALYSIS***


Age Count Age Percentitle Total Purchase  Purchase Count  \
Age Group                                                             
<10               22           3.84%         $96.62              32   
10-14             20           3.49%         $83.79              31   
14-18             84          14.66%        $319.32             111   
18-22            178          31.06%        $676.20             231   
22-26            153          26.70%        $608.02             207   
26-30             44           7.68%        $187.99              63   
30-34             34           5.93%        $141.24              46   
34-38             25           4.36%        $104.06              37   
38-42             11           1.92%         $62.56              20   
42-46              2           0.35%          $6.53               2   

          Average Purchase Normalized Purchase  
Age Group                                       
<10                  $3.02               $4.39  
10-14                $2.70               $4.19  
14-18                $2.88               $3.80  
18-22                $2.93               $3.80  
22-26                $2.94               $3.97  
26-30                $2.98               $4.27  
30-34                $3.07               $4.15  
34-38                $2.81               $4.16  
38-42                $3.13               $5.69  
42-46                $3.26               $3.26

In [98]:
#top spenders
#Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
#SN, purchase count, average purchase price, total purchase value
top_spender = data.groupby(["SN"])
top_sn = pd.DataFrame(top_spender["Price"].sum()).sort_values(["Price"], ascending=False)
top_sn_count = pd.DataFrame(top_spender["Price"].count())
top_sn_avg = round(pd.DataFrame(top_spender["Price"].mean()),2)
top_sn_rename = top_sn.rename(columns={"Price": "Total Purchase"}) 
top_count_rename = top_sn_count.rename(columns={"Price": "Purchase Count"}) 
top_avg_rename = top_sn_avg.rename(columns={"Price": "Average Purchase"}) 
top_sn_rename["Purchase Count"] = top_count_rename
top_sn_rename["Average Purchase"] = top_avg_rename
top_sn_rename["Total Purchase"] = top_sn_rename["Total Purchase"].map("${:.2f}".format)
top_sn_rename["Average Purchase"] = top_sn_rename["Average Purchase"].map("${:.2f}".format)

In [137]:
#final spenders analysis
top_summary = top_sn_rename[["Purchase Count", "Total Purchase", "Average Purchase"]]
print("***MOST SPENDER ANALYSIS***")
top_summary.head()

***MOST SPENDER ANALYSIS***


Purchase Count Total Purchase Average Purchase
SN                                                         
Undirrala66               5         $17.06            $3.41
Saedue76                  4         $13.56            $3.39
Mindimnya67               4         $12.74            $3.18
Haellysu29                3         $12.73            $4.24
Eoda93                    3         $11.58            $3.86

In [100]:
#most popular 5 items, itemID, item name, purchase count, item price, total purchase value
#Upon much research I noticed that someone item has two different IDs ex)item called Final Critic has both item ID 92 and item ID 101.
#I have decided to group them by both name and ID and sort the data based on item I
#item search result for Fianl Critic item. it has two different item IDs
data.loc[data["Item Name"] == "Final Critic", "Item ID"]

3       92
54     101
193    101
226     92
415     92
431     92
458     92
509     92
572     92
606    101
612    101
648     92
663    101
737    101
Name: Item ID, dtype: int64

In [135]:
top_item = data.groupby(["Item Name", "Item ID"])
top_id = pd.DataFrame(top_item["Item ID"].count())
top_id_rename = top_id.rename(columns={"Item ID":"Purchase Count"})
item_total = pd.DataFrame(top_item["Price"].sum())
item_total_rename = item_total.rename(columns={"Price":"Total Price"})
top_id_rename["Total Purchase"] = item_total_rename
item_price = pd.DataFrame(top_item["Price"].sum()/top_item["Item ID"].count())
item_price_rename = item_price.rename(columns={0 :"Item Price"})
top_id_rename["Item Price"] = item_price
top_id_rename.head()

Purchase Count  Total Purchase  \
Item Name                      Item ID                                   
Abyssal Shard                  162                   3            6.12   
Aetherius, Boon of the Blessed 137                   4           19.00   
Agatha                         120                   5            9.55   
Alpha                          130                   7           10.92   
Alpha, Oath of Zeal            79                    7           20.16   

                                        Item Price  
Item Name                      Item ID              
Abyssal Shard                  162            2.04  
Aetherius, Boon of the Blessed 137            4.75  
Agatha                         120            1.91  
Alpha                          130            1.56  
Alpha, Oath of Zeal            79             2.88

In [132]:
#now we are going to sort out the above table to determine the most popular items(by count number) and most profitable items(by total purchase number).
item_popular = top_id_rename.sort_values(["Purchase Count"], ascending=False)
item_popular["Total Purchase"] = item_popular["Total Purchase"].map("${:.2f}".format)
item_popular["Item Price"] = item_popular["Item Price"].map("${:.2f}".format)
print("***MOST POPULAR ITEM***")
item_popular.head()

***MOST POPULAR ITEM***


Purchase Count Total Purchase  \
Item Name                            Item ID                                  
Betrayal, Whisper of Grieving Widows 39                   11         $25.85   
Arcane Gem                           84                   11         $24.53   
Retribution Axe                      34                    9         $37.26   
Trickster                            31                    9         $18.63   
Serenity                             13                    9         $13.41   

                                             Item Price  
Item Name                            Item ID             
Betrayal, Whisper of Grieving Widows 39           $2.35  
Arcane Gem                           84           $2.23  
Retribution Axe                      34           $4.14  
Trickster                            31           $2.07  
Serenity                             13           $1.49

In [133]:
item_profit = top_id_rename.sort_values(["Total Purchase"], ascending=False)
item_profit["Total Purchase"] = item_profit["Total Purchase"].map("${:.2f}".format)
item_profit["Item Price"] = item_profit["Item Price"].map("${:.2f}".format)
print("***MOST PROFITABLE ITEM***")
item_profit.head()

***MOST PROFITABLE ITEM***


Purchase Count Total Purchase Item Price
Item Name                  Item ID                                          
Retribution Axe            34                    9         $37.26      $4.14
Spectral Diamond Doomblade 115                   7         $29.75      $4.25
Orenmir                    32                    6         $29.70      $4.95
Singed Scalpel             103                   6         $29.22      $4.87
Splitter, Foe Of Subtlety  107                   8         $28.88      $3.61

##### DATA ANALYSIS

* Most popular gender group in Heroes of Pymoli is male and age group is 18-22.
* Although most popular item in the game was Betrayal, Whisper of Grieving Widows(ID:39), most profitable item was Retribution Axe(ID:34), which was sold for 4.14 dollars each. 
* Each player spent nearly 3 dollars purchase on items.